### ISSUE TITLE
Create a branch from an issue, with pull/merge request

### ISSUE THREAD

**Issue by ygoe** [go-gitea/gitea#20226]
### Feature Description

There is a workflow that uses issues to define tasks to do and then work branches to do the issue's work in. When the work is done, the branch will be merged into master/main. It saves a lot of manual work if the Git tool already creates a work branch from an issue and also a merge/pull request for it with the corresponding "Closes issue ..." message.

GitLab does this and I'm using it a lot. GitHub also does this but that's not (yet?) reflected in Gitea's [feature comparison](https://docs.gitea.io/en-us/comparison/#issue-tracker).

It has been [asked before](https://github.com/go-gitea/gitea/issues/19827) but was closed. Not sure why because it was neither implemented nor rejected and can't be reopened for some reason. Here's a second try.

### GitHub

![grafik](https://user-images.githubusercontent.com/868315/177098081-e2bfc13d-7ea7-464b-87b6-9a6c589428e0.png)

![grafik](https://user-images.githubusercontent.com/868315/177098167-600f2fdf-2058-4394-9fae-b400c3389b2f.png)

### GitLab

![grafik](https://user-images.githubusercontent.com/868315/177098365-a222e804-48be-46c0-afcd-4c3f295bebc2.png)

![grafik](https://user-images.githubusercontent.com/868315/177098463-f9ad2014-f246-4921-82ad-2624cfc81227.png)
(Clicking the button, not its dropdown menu, would immediately create the branch and MR)

**Comment by flipkickmedia** [comment]
I was just thinking the same thing :)

If we could allow the branch name to be templated that would be a +1 on gitlab/hub

It would be great to see this and it's probably a template/UI change with maybe a setting in the repo settings to provide the template for the branch name.

Also, I wouldn't want to create a branch until the ticket was assigned, so some automation here would be great to see. A repo settings checkbox perhaps to "create issue branch on assignment", "create on label added <dropdown with labels>" etc 

The MVP would be just to add in the create branch option inside the issue and allow for manual branch name.

**Comment by fairking** [comment]
> Also, I wouldn't want to create a branch until the ticket was assigned, 

I would recommend to look at the dev.azure.com. they have done it very nacely.

![Screenshot_20220808-005815](https://user-images.githubusercontent.com/13495631/183316244-1c666e4c-9782-43fe-b9f7-dbbdcd04ffbd.png)

**Comment by smallswan399** [comment]
Anyone who working on this?

**Comment by Taywee** [comment]
@fairking
> BTW. I cannot see why do I need to create a branch and a PR at the same time as Gitlab does. The PR means the work is done and ready for review and approval - while when I create a branch I have no any code changes at all. The reviewer would be confused with many empty or not completed PRs. PR should be created when code changes has been completed and ready for review.

This is just a different development model, and I appreciate it.  The major points about the GitLab model is:

* The PR is opened with a "WIP:" prefix.  There is a conceptual and functional difference between WIP and non-WIP PRs, and you can filter them accordingly.
* The PR has a button that lets you auto-remove the WIP prefix and mark it as "ready".
* When you're creating a work branch, you almost certainly will want a PR anyway eventually, so it's one less action to just get it out of the way at the beginning.
* If you have testing CI that runs on PRs, this lets you more easily work with it as you develop.  I think you can even prevent marking a WIP PR as ready until tests pass, but I'm not certain (might be an EE only feature).
* It lets you more easily see the actual work associated with an issue.  An issue can have multiple branches associated with it, and some people branch while working on a single issue, and might end up uploading those branches, not to mention other release branches and things that aren't associated with specific issue or PR work.  It also lets people more easily leave comments on ongoing work in the PR rather than digging through branches, which can be extra helpful if you happen to have multiple people working on one PR at a time.

From my perspective, if an issue will definitely need at least one branch and at least one PR associated with that branch, it's a productivity gain to just make them both, all linked together and to the issue, with one action.  GitHub and Azure's model involves more effort for no real extra gain in power or flexibility, and Gitea's is all manual at the moment.  I also thought the GitLab way was weird at first, but it does save some busywork.

**Comment by dsxmachina** [comment]
I think the feature would be very nice from people migrating over from gitlab.

I work on a self-hosted gitlab instance in my company, and I am not happy about how slow it runs.
Many developers share this opinion, **but** the above described workflow is something that my company (and all the project managers etc.) are *used to*. 
People like me cannot pitch gitea to those people.

Please have mercy with me and all the others and consider implementing this feature :pray: :pray: :pray: 
Free us from the shackles :chains:  of slow loading issue pages and a bloat interface.

( :wink: )

But srsly though: The feature would be nice IMHO.

**Comment by Taywee** [comment]
@fairking I agree.  This is actually something that I've been handling client-side.  [I have a custom git command called `issue-pr`](https://codeberg.org/Taywee/berx) (warning: very rough-and-ready undocumented alpha-quality code) that uses the gitea API to do the same process client side.  Running `git issue-pr $issue_number` will create a branch for the issue and a WIP PR and link them all together.

A browser extension or userscript could probably do the same thing, now that I think of it.

**Comment by orange-pig** [comment]
An obvious requirement is to create a branch from an issue. The PR does need to be created after the branch function is completed. Perhaps an additional function to create a PR on the current issue can be provided.

**Comment by Dav-11** [comment]
hello, is this feature still being developed ?

**Comment by lunny** [comment]
Nobody are working on this issue.

**Comment by lunny** [comment]
This is not PR first and issue later. It means, you can create a new issue from an issue/pr/comments if you find it related to another problem but you don't want to discuss/resolve in the same place. And that's another feature that is not the same as the current.

And I think your suggestion is great to require an issue before sending a PR.

**Comment by janstieler** [comment]
Hi,
I worked in the last half year on a massive software project and they used Gitlab with a workflow like the threadopener mentioned. And it worked really well. In my own I have smaller projects but want also use this workflow. So I installed gutes because it had not so the overhead like gitlab. Gitlab had a lot of functions I don't need.
But yesterday I had find out that this perfect workflow us not possible in Gitea and I can't it understand because it is a perfect process workflow with a logic base.

So will it come or should I use Gitlab?

Best

**Comment by lunny** [comment]
> Hi, I worked in the last half year on a massive software project and they used Gitlab with a workflow like the threadopener mentioned. And it worked really well. In my own I have smaller projects but want also use this workflow. So I installed gutes because it had not so the overhead like gitlab. Gitlab had a lot of functions I don't need. But yesterday I had find out that this perfect workflow us not possible in Gitea and I can't it understand because it is a perfect process workflow with a logic base.
> 
> So will it come or should I use Gitlab?
> 
> Best

I'm interested in sending a PR to implement that.

**Comment by lunny** [comment]
#31899 supports creating branches and pull requests from the new sidebar section `development`. Linking branches and pull requests will be the next step.
